contributors/design-proposals/auth/pod-restriction.md

```diff
@@ -0,0 +1,477 @@
+    NodeSelector, etc.
+
+### Policy Matching - Union or Intersection?
+
+**Intersection** refers to the model where a Pod must be valid under all PodRestrictions that
+pertain to it. The advantage of intersecting is that restrictions can be composed. For example, I
+could have a policy that frontend & backend Pods cannot share a node, and a separate policy that all
+production Pods must drop privileges. Without the ability to intersect those policies, I need 3
+different versions (either, or, both). A problem with this approach is that running a privileged
+namespace (e.g. kube-system) means that every restriction needs an exception in the
+NamespaceSelector.
+
+**Union** refers to the model where a Pod must be valid under any PodRestriction that is bound to
+it. This approach makes it easier to reason about what the restrictions on a given namespace are,
+and easier to add exceptions for namespaces, but can lead to a combinatorial explosion of different
+policies that can become unmanageable.
+
+**Proposal**: Why not both? The default behavior is intersection, but you can punch a hole through
+for a specific namespace with a union approach. An optional `BindingMode` is added to the Binding
+which specifies the behavior. A pod must be valid under either:
+
+-   ANY `BindingModeAccept` policy that is bound to it, or
+
+-   ALL `BindingModeDrop` policies that are bound to it
+
+In other words, the Pod is first checked against Accept restrictions, and is accepted if any
+match. Then, it is checked against the Drop restrictions, and rejected if any don't match. The
+default behavior at the end is to accept the pod.
+
+```
+Binding:
+    Mode BindingMode
+
+BindingModeAccept
+BindingModeDrop
+```
+
+### Admission Handling
+
+The PodRestriction controller handles all create, update, and patch requests for Pod resources. It
+is also applied to all (native) resources that create pods via a PodTemplate, including ReplicaSets,
+ReplicationControllers, DaemonSets, StatefulSets, Deployments, Jobs, etc.
+
+#### Justification
+
+From a security perspective, it should be sufficient to apply the restriction to Pod requests, so
+why validate other objects? Fail-fast. By surfacing an error before the controller object is
+created, the user gets immediate feedback that they are attempting a forbidden operation. The
+alternative experience is that the controller object creation succeeds, but then the user must
+follow debug steps to determine why Pods are not being created.
+
+#### Optional Extensions
+
+New CRDs that hold a PodTemplate and create Pods should be able to be registered against the
+PodRestriction controller. The registration should include a path (auto-discovery?) to extract the
+PodTemplate from the CRD.
+
+### API Design
+
+#### Monolith vs. Modular
+
+Our current approach to policies has been "modular", i.e. a different controller and API for each
+"area" of features. This has led to some inconsistencies and gaps in our policies, and makes adding
```

Comment:
I think we have gone past that stage of "area" of features as well. While I am not completely against monolithic approach after your point, I think, having a delegation model to area would helpful and might be a good step. For example, podrestriction controller would delegate to scheduling admission controller which can add restrictions specific to scheduling. But a careful grouping of controllers to areas is very much needed.